Header stream tests

jvkersch opened the pull request:
Tests + documentation for https://github.com/EGA-archive/crypt4gh/tree/header-stream.

silverdaz commented on docs/usage.rst:
```diff
@@ -64,3 +64,27 @@ Any user can generate a keypair with:
+
+.. code-block:: console
+
+    $ cat header.bob.c4gh M.data.c4gh > M.c4gh
```
Use `cat header.bob.c4gh M.data.c4gh | crypt4gh decrypt --sk bob.sec > M`. No need to create a new file.

jvkersch replied:
Done!

silverdaz commented on tests/header_stream.bats:
```diff
@@ -0,0 +1,60 @@
+    crypt4gh encrypt --sk ${BOB_SECKEY} --recipient_pk ${BOB_PUBKEY} --header $TESTFILES/header.bob.c4gh < $TESTFILE > $TESTFILES/data.c4gh
+
+    # Bob changes the header for Alice
+    crypt4gh reencrypt --sk ${BOB_SECKEY} --recipient_pk ${ALICE_PUBKEY} < $TESTFILES/header.bob.c4gh > $TESTFILES/header.alice.c4gh
```
That works because the payload in the header-only file is empty, right? That's a strange side effect, I think.

jvkersch replied:
That is correct. The re-encrypt step just copies over the payload, so if there is none, it does nothing. I agree that it is somewhat unexpected, and brittle too (if a future version of the code changes how the payload is handled, then this might break).

silverdaz commented on tests/header_stream.bats:
```diff
@@ -0,0 +1,60 @@
+    crypt4gh reencrypt --sk ${BOB_SECKEY} --recipient_pk ${ALICE_PUBKEY} < $TESTFILES/header.bob.c4gh > $TESTFILES/header.alice.c4gh
+
+    # Alice concatenates the header and data
+    cat $TESTFILES/header.alice.c4gh $TESTFILES/data.c4gh > $TESTFILES/message.c4gh
```
Use a pipe, no need to create a separate file. You might not have the space for it.

silverdaz reviewed (commented):
I think I see what you mean with "reencrypt for free", but it's a rather strange side-effect of not carrying any payload... Not sure it should be intended to be used that way...

jvkersch commented:
@silverdaz Thanks for the quick follow-up. I've addressed your comment regarding the temporary files.

For the re-encryption scenario, I would like to have some functionality so that I can create a new header for a different user. I think this fits the use case of storing the encrypted payload somewhere, and only (re-)creating a header, very well. But I agree that it is unexpected that this is supported as-is by the reencrypt command. It is also likely to break if in a future version of the code the payload is not simply copied over, but handled in a different way.

I don't know what the best way to handle this is, but in my last commit (9bdc802) I added an extra option `--header-only` to the re-encrypt command to indicate that re-encrypt should only create a new header, not copy over the payload. I am not sure that this is the best solution, but at least it has the advantage of being explicit. Happy to hear your thoughts!

silverdaz commented on crypt4gh/lib.py:
```diff
@@ -416,13 +416,14 @@ def reencrypt(keys, recipient_keys, infile, outfile, chunk_size=4096, trim=False
-            break
-        outfile.write(data)
+    # Stream the remainder, if present
+    if not header_only:
```
Good idea.
Make it differently though:
```
if header_only:
    LOG.info('Header-only reencryption Successful')
     return
...
the rest
```

silverdaz reviewed (commented):
header-only sounds like a good idea 👍

jvkersch commented:
Thanks, returning early in the case of header-only reencryption is a good idea. I've made that change, and I've also cleaned up one instance in the docs where I was creating an unnecessary temp file.

silverdaz reviewed (approved):
LGTM
